tests/test_controls.py

```diff
@@ -0,0 +1,494 @@
+        assert isinstance(getattr(self.base_procedure, property), var_type)
+    
+    @pytest.mark.parametrize("property, value", [('parallel', ParallelOptions.All),
+                                                 ('calcSldDuringFit', True),
+                                                 ('resamPars', [0.2, 1]),
+                                                 ('display', DisplayOptions.Notify)])
+    def test_base_procedure_setters(self, property: str,  value: Any) -> None:
+        setattr(self.base_procedure, property, value)
+        assert getattr(self.base_procedure, property) == value
+    
+    @pytest.mark.parametrize("var, exp_type", [('string', float),
+                                               (True, str),
+                                               (1, bool),
+                                               (1.0, int)])
+    def test_validate_type(self, var: Any, exp_type: Any) -> None:
+        with pytest.raises(TypeError) as exc:
+            self.base_procedure._validate_type('var', var, exp_type)
+        assert f"var must be of type {exp_type.__name__}" in str(exc.value)
+
+    @pytest.mark.parametrize("enum, enum_type", [(ParallelOptions, str),
+                                                 (Procedures, str),
+                                                 (DisplayOptions, str),
+                                                 (BoundHandlingOptions, str),
+                                                 (StrategyOptions, int)])
+    def test_validate_value(self, enum: str, enum_type: Union[int, str]) -> None:
+        allowed_options = [str(o.value) for o in enum]
+        with pytest.raises(ValueError) as exc:
+            self.base_procedure._validate_value('var', 'test_variable', enum, enum_type)
+        assert (f"var must be a {enum.__name__} "
+        f"enum or one of the following {enum_type.__name__} "
+        f"{', '.join(allowed_options)}") == str(exc.value)
+
+    @pytest.mark.parametrize("lower, upper, msg", 
+                             [(True, True, "var must be greater than 0 and less than 1"),
+                              (False, True, "var must be greater than or equal to 0 and less than 1"),
+                              (True, False, "var must be greater than 0 and less than or equal to 1"),
+                              (False, False, "var must be greater than or equal to 0 and less than or equal to 1")])
+    def test_validate_range(self, lower: bool, upper: bool, msg: str) -> None:
+        with pytest.raises(ValueError) as exc:
+            self.base_procedure._validate_range(name='var',
+                                                value=10,
+                                                lower_limit=0,
+                                                upper_limit=1,
+                                                lower_exclusive=lower,
+                                                upper_exclusive=upper)
+        assert msg == str(exc.value)
```

Comment:
You can also use the `match` option of `pytest.raises` to include the error message within the context manager. The syntax is `with pytest.raises(ValueError, match=msg) as exc:`

Reply:
I think comparing the messages using assert is better than using the context manager. I tried it with the context manager. It is just checking if `msg` is in the error message. It is using `regex`. If you pass `" "` (a str with space), then the tests pass. To test the entire string might be better.

Reply:
Hmm, that doesn't seem to work. Never mind . . .